### ISSUE TITLE
Tags support

### ISSUE THREAD

**Issue by spakka** [RIPE-NCC/ripe-atlas-tools#25]
Hiya,

If I want to do a v6 ping from many atlas probes, lets say I'm using the defaults of 50 from region WW,
My default expectation is that it should not select probes where v6 doesn't work.
Since the system has 'system-ipv6-works' tags assigned to probes, I'd like to be able to set tags in the measurement creation.

For extra bonus points, a v4 measurement should only select 'system-ipv4-works' and v6 measurements should only select 'system-ipv6-works' by default (or allow me to configure it in my defaults)

That'd be awesome and shiny :-)

Cheers,
Colin

**Comment by ioc32** [comment]
:+1: to both the ability to set arbitrary tags for probe selection as well as sane defaults (system-ipv{4,6}-works, for operators ;-)

**Comment by pierky** [comment]
This is my idea (717e65f50203f31c344f8b2530137976a6e3f800):

* tags can be passed via CLI in the **measure** command, using the **--include-tag** and **--exclude-tag** arguments (even multiple times to build a list of include/exclude tags);
* if no tags are supplied via CLI, they are read from the configuration:
  * first, the format *msm_type*.*addr_family* is tried, to allow default tags specific for the current measurement type and address family (example: dns.4);
  * then, only `af`.*addr_family* is tried, for example to store the *system-ipv6-works* and *system-ipv4-works* default tags.

I thought to put only *system-ipv6-works* and *system-ipv4-works* at `af` level within the default config:

```YAML
# Tags added to probes selection
smart-tags:
  af:
    '4':
      exclude: []
      include:
      - system-ipv4-works
    '6':
      exclude: []
      include:
      - system-ipv6-works
```

So, every measurement would have its *system-ipv[4|6]-works* tag automatically added, depending on the requested address family:

```
$ ripe-atlas measure ping --target www.ripe.net --af 6 --dry-run
Definitions:
...
Sources:
  requested: 50
  type: area
  value: WW
  tags: {'exclude': [], 'include': ['system-ipv6-works']}
```


Optionally, users can add their most-specific preferred tags in their config file:

```YAML
# Tags added to probes selection
smart-tags:
  dns:
    '4':
      exclude: []
      include:
      - system-ipv4-works
      - system-can-resolve-a

```

```
$ ripe-atlas measure dns --query-argument www.ripe.net --dry-run
Definitions:
...
Sources:
  requested: 50
  type: area
  value: WW
  tags: {'exclude': [], 'include': ['system-ipv4-works', 'system-can-resolve-a']}

```

**Comment by danielquinn** [comment]
Ah I see now.  I'd like to make a minor modification though, unless you have objections.

Instead of a structure like this:

```yaml
smart-tags:
    af:
        "4":
            exclude: []
            include:
                - system-ipv4-works
        "6":
            exclude: []
            include:
                - system-ipv6-works
    dns:
        "4":
            exclude: []
            include:
                - system-ipv4-works
                - system-can-resolve-a
        "6":
            exclude: []
            include:
                - system-ipv6-works
    ping:
        ...
```

I propose one like this:

```yaml
tags:
    ipv4:
        ping:
            ...
        dns:
            exclude: []
            include:
                - system-can-resolve-a
        all:
            exclude: []
            include:
                - system-ipv4-works
    ipv6:
        ping:
            ...
        dns:
            exclude: []
            include:
                - system-can-resolve-aaaa
        all:
            exclude: []
            include:
                - system-ipv6-works
```

The logic being that of a fall-through-like interface:  If your measurement is `v4` and `dns`, then the tags will be collected from `tags.ipv4.dns` and `tags.ipv4.all`.  I'd also want defaults for all types in the `Configuration` class, not just the ones we've got right now.

How's that sound by you?

**Comment by danielquinn** [comment]
I'd also like to put it under `specification.source.tags` rather than top-level as you have it currently.

**Comment by danielquinn** [comment]
Strike that last part, I'm not really sure where I want it to live yet.  On the one hand, it doesn't belong in the root because it's only used for measurement creation, but it doesn't belong in `specification` because it doesn't conform to the way you'd typically specify a measurement.  I'm going to have to think a bit on this part.

**Comment by pierky** [comment]
:+1:  to your structure.
I tried to put it into `specification` too but it conflicted with msm creation and I preferred to take it off.

**Comment by danielquinn** [comment]
Alright then.  After the two big merges we did today, your PR is in conflict, so I've started tinkering with it.  I've merged all of your code and now I'm working with the structure change as well as adding in some af-guessing.  When I'm done, I'll post another PR (from your original) and link to it here.  At that point I'd appreciate a :+1: from you just to see if I've done what you needed, and then we'll merge it if all is good.
